docs/rfc/007-endpoint-errors.md

```diff
@@ -0,0 +1,212 @@
+endpoints fit nicely into that. Further, it is a well known language feature that requires no external libraries and 
+clearly shows the lineage of errors that are thrown.
+
+**Enriched Types**
+
+With an expressive type system, it is possible to express the different possible outcomes as an enriched version of the
+normal return type. This could be done, for example, by using union types, which already exist in the Conjure universe.
+Some languages are more amiable to union types than others.
+
+**Why The Demo Uses Checked Exceptions**
+
+As of now, `SerializableError` does not lend itself to complex data types as complex error objects are flattened into a 
+`Map<String, String>`, which makes parsing it out very difficult. Checked exceptions preserve the existing status-quo 
+(that is, not messing with the reality of  `SerializableError`), but still improve expressivity and enforce errors being 
+actually handled. 
+
+Further to this, leveraging checked exceptions provides a few other benefits - it is an existing language feature that
+is well-supported by all IDEs, and it makes it simpler to find exactly where a particular error case is coming from (by
+following the exception declarations).
+
+Conversely, a `Result` type (or similar) would involve constructing our own library (or importing an external one), is 
+not currently ergonomically supported by the Java language (until we get something resembling `switch` statements for 
+sealed classes), and would require error-prone rules to enforce that error cases are actually thrown (rather than just
+ignored in the error case).
+
+### Example Server Implementations
+
+**Undertow Service**
+
+```
+@Generated("com.palantir.conjure.java.services.UndertowServiceInterfaceGenerator")
+public interface MyConjureService {
+    /**
+     * Returns some result, but once in a blue moon throws an error
+     * @apiNote {@code GET /}
+     */
+    List<ResourceIdentifier> getSomeResult(AuthHeader authHeader)
+            throws MyConjureServiceErrors.BlueMoonServiceException;
+```
+
+**Undertow Resource**
+
+```
+public final class MyConjureResource implements MyConjureService {
+
+   ...
+
+   @Override
+   public List<ResourceIdentifier> getSomeResult(AuthHeader authHeader)
+          throws BlueMoonServiceException {
+      throw MyConjureServiceErrors.throwIfBlueMoon(isBlueMoon);
+   }
+```
+
+### Example Dialogue Client Implementation via Exceptions
+
+
+```
+MyConjureServiceBlocking myConjureService = 
+    witchcraft.conjureClients()
+        .client(MyConjureServiceBlocking.class, "my-conjure-service").get();
+
+try {
+    myConjureService.getSomeResult(AuthHeader.valueOf("authHeader"));
+} catch (BlueMoonRemoteException e) { // <- checked exception so has to be caught!
+    handleBlueMoonEvent();
+}
+```
+
+## Open Questions ##
+
+1) Should handling errors on endpoints be mandatory? While it seems like a good idea in principle, could it cause 'bad'
+   behavior in client or server code?
```

Comment:
If I'm reading this correctly, this feels like a java-specific question around checked vs. unchecked exceptions. There's no idiomatic equivalent to a checked exception in C# or python (i.e., all errors "optional" in those languages). On the other hand, in languages like golang or rust, all error handling is "mandatory" by design. Basically, I'm not sure the mandatory vs optional question makes sense it the broader ecosystem of other languages.

Reply:
I vote that language ergonomics / idiomatic-ness is more important than trying to force every language to adopt a uniform error-checking mechanism (e.g., a result type packed within a union type that supports errors).

Reply:
+1 to @pt-achang 's point